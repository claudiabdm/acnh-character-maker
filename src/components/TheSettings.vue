<template>
  <section class="settings">
    <nav class="settings__buttons" aria-label="Settings options">
      <router-link :class="[
        'settings__link-wrapper',
        { 'settings__link-wrapper--active': selectedTab === icon },
        'settings__link'
      ]" v-for="icon in icons" :key="icon" :aria-label="icon === 'nose-mouth' ? icon.replace('-', ' / ') : icon"
        :to="{ path: icon, query: { ...$route.query } }">
        <svg viewBow="0 0 1 1" class="settings__link-icon">
          <use :href="iconPath(icon)" />
        </svg>
      </router-link>
    </nav>
    <div class="settings__options">
      <router-view v-slot="{ Component }">
        <transition name="fade" mode="out-in">
          <component :is="Component" />
        </transition>
      </router-view>
    </div>
  </section>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import icons from '/icons.svg?url';

export default defineComponent({
  data() {
    return {
      icons: [
        'skin',
        'hair',
        'eyes',
        'nose-mouth',
        'blush',
        'background',
        'clothes'
      ],
      selectedTab: this.$route.name || 'skin'
    };
  },
  watch: {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    $route(to, _from) {
      if (this.selectedTab !== to.name) {
        this.onSelectTab(to.name);
      }
    }
  },
  methods: {
    iconPath(icon: string): string {
      return icons + '#' + icon;
    },
    onSelectTab(icon: string): void {
      console.log(icon)
      this.selectedTab = icon;
    }
  }
});
</script>

<style lang="scss" scoped>
@use '@/styles/config.scss' as *;

.settings {
  @include size(100%, 100%);
  min-height: 100vh;
  background-color: var(--primary-100);
  overflow: hidden;

  &__options {
    @media screen and (min-width: 1024px) {
      height: calc(100% - 135px);
      overflow-y: auto;
    }
  }

  &__buttons {
    @include size(100%, max-content);
    @include flex(center, center);
    flex-wrap: nowrap;
    padding: rem($padding-s + 10px);
  }

  &__link-wrapper {
    @include flex(center, center);
    position: relative;
    border-radius: 50%;
    background-color: var(--tertiary-200);
    color: var(--base);
    font-size: 1rem;
    margin: 0;
    margin-left: -10px;
    padding: 0;

    &::before {
      content: '';
      @include size(0, 0);
      opacity: 0;
      position: absolute;
      border-radius: rem(20px);
      transform: translate3d(0, -100%, 0);
      transition: transform 0.25s cubic-bezier(0.175, 0.885, 0.32, 1.275);
      z-index: 1;
    }

    &--active {
      .settings__link-icon {
        color: var(--base);
      }

      &::before {
        @include size(max-content, auto);
        @include flex(center, center);
        content: '   ' attr(aria-label) '   ';
        background-color: var(--tertiary-100);
        color: var(--tertiary-200);
        padding: rem(5px) 0;
        text-transform: capitalize;
        font-size: $text-xs;
        transform: translate3d(0, -130%, 0);
        opacity: 1;
        white-space: pre;

        @media screen and (min-width: 768px) {
          font-size: 1rem;
        }
      }
    }
  }

  &__link {
    @include flex(center, center);
    @include size(rem(60px), rem(60px));

    @media screen and (min-width: 768px) {
      @include size(rem(70px), rem(70px));
    }
  }

  &__link-icon {
    @include size(50%, 50%);
    color: var(--tertiary-100);
    pointer-events: all;
    transition: color 0.15s;
  }
}
</style>
