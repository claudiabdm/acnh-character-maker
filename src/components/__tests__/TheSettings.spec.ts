/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { flushPromises, shallowMount, VueWrapper } from "@vue/test-utils";
import TheSettings from "../TheSettings.vue";
import { createRouter, createWebHistory } from "vue-router";

describe("The Settings", () => {
  const $route = {
    query: {
      skin: "f0a06f",
      hair: "princess",
      hairColor: "fff",
      eyes: "eyes-oval",
      eyesColor: "fff",
      nose: "nose-oval",
      mouth: "mouth-cat",
      blush: "solid-coral",
      backgroundColor: "000",
      clothes: "tee-2",
    } as { [key: string]: string },
  };

  let wrapper: VueWrapper<any>;

  beforeEach(() => {
    wrapper = shallowMount(TheSettings, {
      global: {
        mocks: {
          $route,
        },
        stubs: ["router-link", "router-view"],
      },
    });
  });
  it("should match the snapshot", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should change selected tab if it was not selected when route changed", async () => {
    const router = createRouter({
      history: createWebHistory(),
      routes: [
        {
          path: "/",
          component: TheSettings,
          children: [
            {
              path: "/skin",
              name: "skin",
              component: {
                template: "skin",
              },
            },
            {
              path: "/eyes",
              name: "eyes",
              component: {
                template: "eyes",
              },
            },
          ],
        },
      ],
    });
    const wrapperRouter = shallowMount(TheSettings, {
      global: { plugins: [router] },
    });
    router.push("/skin");
    await router.isReady();
    expect(wrapperRouter.vm.selectedTab).toBe("skin");

    router.push("/eyes");
    await router.isReady();
    await flushPromises();
    expect(wrapperRouter.vm.selectedTab).toBe("eyes");
  });

  it("should return path to svg icon", () => {
    const path = wrapper.vm.iconPath("skin");
    expect(path).toBe("/icons.svg#skin");
  });

  it("should change selected tab", () => {
    wrapper.vm.onSelectTab("eyes");
    expect(wrapper.vm.selectedTab).toBe("eyes");
  });
});
