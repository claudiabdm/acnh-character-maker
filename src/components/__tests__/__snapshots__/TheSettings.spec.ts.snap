// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The Settings > should match the snapshot 1`] = `
<section
  class="settings"
  data-v-9b072b1e=""
>
  <nav
    aria-label="Settings options"
    class="settings__buttons"
    data-v-9b072b1e=""
  >
    
    <button
      aria-label="skin"
      class="settings__link-wrapper settings__link-wrapper--active"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to skin tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    <button
      aria-label="hair"
      class="settings__link-wrapper"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to hair tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    <button
      aria-label="eyes"
      class="settings__link-wrapper"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to eyes tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    <button
      aria-label="nose / mouth"
      class="settings__link-wrapper"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to nose-mouth tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    <button
      aria-label="blush"
      class="settings__link-wrapper"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to blush tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    <button
      aria-label="background"
      class="settings__link-wrapper"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to background tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    <button
      aria-label="clothes"
      class="settings__link-wrapper"
      data-v-9b072b1e=""
    >
      <router-link-stub
        aria-label="Go to clothes tab"
        class="settings__link"
        data-v-9b072b1e=""
        to="[object Object]"
      />
    </button>
    
  </nav>
  <div
    class="settings__options"
    data-v-9b072b1e=""
  >
    <router-view-stub
      data-v-9b072b1e=""
    />
  </div>
</section>
`;
